File: OOp.py
#........................................polymorphism................
#Overloading
# class Calculator:
#     def add(self, a, b=None, c=None):
#         if b is not None and c is not None:
#             return a + b + c
#         elif b is not None:
#             return a + b
#         else:
#             return a
#         """
#         ..it is like..
#         def add(self, a):
#         return a
#         def add(self, a,b):
#         return a+b
#         def add(self, a,b,c):
#         return a+b+c
#         """
#
# # Creating object
# calc = Calculator()
#
# # Method overloading-like behavior
# print(calc.add(1))      # Output: 1
# print(calc.add(1, 2))   # Output: 3
# print(calc.add(1, 2, 3)) # Output: 6




#Overriding
# class Animal:
#     def sound(self):
#         print("Animal makes a sound")
#
# class Dog(Animal):
#     def sound(self):
#         print("Dog barks")
#
# # Creating objects
# animal_obj = Animal()
# dog_obj = Dog()
#
# # Method overriding in action
# animal_obj.sound()  # Output: Animal makes a sound
# dog_obj.sound()     # Output: Dog barks




#simple
# class Animal:
#     def make_sound(self):
#         pass
#
# class Dog(Animal):
#     def make_sound(self):
#         return "Woof!"
#
# class Cat(Animal):
#     def make_sound(self):
#         return "Meow!"
#
# # Polymorphic behavior
# animals = [Dog(), Cat()]
#
# for animal in animals:
#     print(animal.make_sound())


#.......................Abstract class...............................

# #Abstract Class
# from abc import ABC, abstractmethod
#
# # Abstract class declaration
# class Animal(ABC):
#     def __init__(self, name):
#         self.name = name
#
#     @abstractmethod
#     def make_sound(self):
#         pass  # Abstract method to be implemented by subclasses
#
# # Concrete subclass of the abstract class
# class Dog(Animal):
#     def make_sound(self):
#         return "Woof!"
#
# # Concrete subclass of the abstract class
# class Cat(Animal):
#     def make_sound(self):
#         return "Meow!"
#
# # Objects and usage
# dog = Dog("Buddy")
# cat = Cat("Whiskers")
#
# print(dog.make_sound())  # Output: Woof!
# print(cat.make_sound())  # Output: Meow!


# #interface of abstract
# from abc import ABC, abstractmethod
#
# # Interface declaration using ABC
# class Shape(ABC):
#     @abstractmethod
#     def calculate_area(self):
#         pass  # Abstract method to be implemented by subclasses
#
# # Class implementing the interface
# class Circle(Shape):
#     def __init__(self, radius):
#         self.radius = radius
#
#     def calculate_area(self):
#         return 3.14 * self.radius ** 2
#
# # Class implementing the interface
# class Square(Shape):
#     def __init__(self, side_length):
#         self.side_length = side_length
#
#     def calculate_area(self):
#         return self.side_length ** 2
#
# # Objects and usage
# circle = Circle(5)
# square = Square(4)
#
# print("Area of Circle:", circle.calculate_area())  # Output: Area of Circle: 78.5
# print("Area of Square:", square.calculate_area())  # Output: Area of Square: 16
#


##simple abstract
# from abc import ABC, abstractmethod
# import math
#
# # Abstract class Shape
# class Shape(ABC):
#     def __init__(self, name):
#         self.name = name
#
#     @abstractmethod
#     def calculate_area(self):
#         pass  # Abstract method to be implemented by subclasses
#
# # Concrete subclass Circle
# class Circle(Shape):
#     def __init__(self, name, radius):
#         super().__init__(name)
#         self.radius = radius
#
#     def calculate_area(self):
#         return math.pi * self.radius ** 2
#
# # Concrete subclass Square
# class Square(Shape):
#     def __init__(self, name, side_length):
#         super().__init__(name)
#         self.side_length = side_length
#
#     def calculate_area(self):
#         return self.side_length ** 2
#
# # Using the abstraction
# circle = Circle("Circle", 5)
# square = Square("Square", 4)
#
# print(f"Area of {circle.name} with radius {circle.radius}: {circle.calculate_area()}")
# # Output: Area of Circle with radius 5: 78.53981633974483
#
# print(f"Area of {square.name} with side length {square.side_length}: {square.calculate_area()}")
# # Output: Area of Square with side length 4: 16




#....................................encaptulation............................

##public
# class MyClass:
#     def __init__(self):
#         self.public_variable = 10
#
# obj = MyClass()
# print(obj.public_variable)  # Output: 10
# obj.public_variable = 20
# print(obj.public_variable)  # Output: 20
#
#
# #private
# class MyClass:
#     def __init__(self):
#         self.__private_variable = 10
#
# obj = MyClass()
# # Attempting to access the private variable directly will result in an error.
# # print(obj.__private_variable)  # Error: 'MyClass' object has no attribute '__private_variable'
#
# # You can access and modify private variables using name mangling.
# print(obj._MyClass__private_variable)  # Output: 10
# obj._MyClass__private_variable = 20
# print(obj._MyClass__private_variable)  # Output: 20
#
#
# #protected
# class MyClass:
#     def __init__(self):
#         self._protected_variable = 10
#
# obj = MyClass()
# print(obj._protected_variable)  # Output: 10
# obj._protected_variable = 20
# print(obj._protected_variable)  # Output: 20
#
#





#...................................Inheritance...........................

# #hierarchical
# class Baba:
#     tk=1000
#     def boka(self):
#         pass
#
# class Cele(Baba):
#     tk = 5
#     def boka(self):
#         return 'Ja porte bos'
#
# class Meye (Baba):
#     tk = 950
#     def boka(self):
#         return 'jao mamoni porte boso'
#
# efat=Cele()
# print(efat.boka())
# print(efat.tk)
#
# sumi=Meye()
# print(sumi.boka())
# print(sumi.tk)
#
# if efat.tk>sumi.tk:
#     print('Efat ar tk bro')
# else:
#     print("sumir tk bro")
#
#
#
#
#



#hybrid
# class Dada:
#     car_val=1000
#     def car(self):
#         return f'dadar car'
#
# class Nana:
#     nanar_tk=10000
#     def tk(self):
#         return f'nanar tk'
#
# class Baba(Dada,Nana):
#     jomi=100000
#     pass
#
# class Ma:
#     mar_tk=10000
#     def ma_ar_tk(self):
#         return f'ma ar tk'
#
#
# class Son(Baba,Ma):
#       pass
#
#
#
# efat=Son()
# print(efat.car())
# print(efat.ma_ar_tk())
# print(efat.tk())
# print(efat.mar_tk + efat.nanar_tk + efat.car_val+efat.jomi)




#Multiple
# class Dada:
#     def driving(self):
#         return f'dadar sompod'
#
# class Nana:
#     def moeny_bag(self):
#         return f'nanar sompod'
#
# class Baba(Dada,Nana):
#     pass
#
#
#
# father=Baba()
# print(father.driving())
# print(father.moeny_bag())
#



#multilevel
# class Gp:
#     def __init__(self,car):
#         self.car=car
#     def driving(self):
#         pass
#
# class Father(Gp):
#     def driving(self):
#         return f'father is driving the {self.car}'
#
# class Son(Father):
#     def driving(self):
#         return f'Son is driving the {self.car}'
#
# son=Son('BMW')
# print(son.driving())




##simple
# class Gp:
#     def __init__(self,car):
#         self.car=car
#     def driving(self):
#         pass
#
# class Father(Gp):
#     def driving(self):
#         return f'father is driving the {self.car}'
#
#
# father=Father('BMW')
# print(father.driving())


##super()
# class Animal:
#     def speak(self):
#         return "Animal speaks"

# class Dog(Animal):
#     def speak(self):
#         # Calling the speak method of the superclass (Animal)
#         # using the super() keyword
#         return super().speak() + " and barks"

# # Usage
# dog = Dog()
# print(dog.speak())  

File: oop.cpp
//--------------------------- Encapsulation-----------------------------

//public-----------
// #include <bits/stdc++.h>
// using namespace std;

// class MyClass {
// public:
//     int myPublicVar;

//     void setMyVar(int value) {
//         myPublicVar = value;
//     }

//     int getMyVar() {
//         return myPublicVar;
//     }
// };

// int main() {
//     MyClass obj;
//     obj.setMyVar(42);

//     cout << "Value of myPublicVar from outside the class: " << obj.myPublicVar << endl;
//     cout << "Value of myPublicVar using member function: " << obj.getMyVar() << endl;

//     obj.myPublicVar = 10;  // Modifying public variable directly

//     cout << "Modified value of myPublicVar: " << obj.myPublicVar << endl;
//     return 0;
// }


//privet---------------
// #include <bits/stdc++.h>
// using namespace std;

// class MyClass {
// private:
//     int myPrivateVar;

// public:
//     void setMyVar(int value) {
//         myPrivateVar = value;
//     }

//     int getMyVar() {
//         return myPrivateVar;
//     }
// };

// int main() {
//     MyClass obj;
//     // obj.myPrivateVar = 42; // This line would cause a compilation error

//     obj.setMyVar(42); // Accessing private variable via public member function

//     cout << "Value of myPrivateVar using member function: " << obj.getMyVar() << endl;

//     return 0;
// }


// //protected-------------
// #include <bits/stdc++.h>
// using namespace std;

// class MyBaseClass {
// protected:
//     int myProtectedVar;

// public:
//     void setMyVar(int value) {
//         myProtectedVar = value;
//     }

//     int getMyVar() {
//         return myProtectedVar;
//     }
// };

// class MyDerivedClass : public MyBaseClass {
// public:
//     void accessProtectedVar() {
//         myProtectedVar = 42; // Accessing protected variable from derived class
//     }
// };

// int main() {
//     MyDerivedClass obj;
//     // obj.myProtectedVar = 42; // This line would cause a compilation error

//     obj.setMyVar(42); // Accessing protected variable via public member function

//     cout << "Value of myProtectedVar using member function: " << obj.getMyVar() << endl;

//     obj.accessProtectedVar(); // Accessing protected variable from derived class

//     return 0;
// }




//--------------------------------Inheritance----------------------------
//single 
// #include <bits/stdc++.h>
// using namespace std;

// class Animal {
// public:
//     void speak() {
//         cout << "Animal speaks" << endl;
//     }
// };

// class Dog : public Animal {
// public:
//     void bark() {
//         cout << "Dog barks" << endl;
//     }
// };

// int main() {
//     Dog dog;
//     dog.speak();  // Output: "Animal speaks"
//     dog.bark();   // Output: "Dog barks"
//     return 0;
// }


////Multilevel Inheritance
// #include <bits/stdc++.h>
// using namespace std;

// class Animal {
// public:
//     void speak() {
//         cout << "Animal speaks" << endl;
//     }
// };

// class Mammal : public Animal {
// };

// class Dog : public Mammal {
// public:
//     void bark() {
//         cout << "Dog barks" << endl;
//     }
// };

// int main() {
//     Dog dog;
//     dog.speak();  // Output: "Animal speaks"
//     dog.bark();   // Output: "Dog barks"
//     return 0;
// }



////Multiple Inheritance
// #include <bits/stdc++.h>
// using namespace std;

// class A {
// public:
//     void methodA() {
//         cout << "Method A" << endl;
//     }
// };

// class B {
// public:
//     void methodB() {
//         cout << "Method B" << endl;
//     }
// };

// class C : public A, public B {
// };

// int main() {
//     C obj;
//     obj.methodA();  // Output: "Method A"
//     obj.methodB();  // Output: "Method B"
//     return 0;
// }


////Hierarchical Inheritance
// #include <bits/stdc++.h>
// using namespace std;

// class Animal {
// public:
//     void speak() {
//         cout << "Animal speaks" << endl;
//     }
// };

// class Dog : public Animal {
// public:
//     void bark() {
//         cout << "Dog barks" << endl;
//     }
// };

// class Cat : public Animal {
// public:
//     void meow() {
//         cout << "Cat meows" << endl;
//     }
// };

// int main() {
//     Dog dog;
//     Cat cat;
//     dog.speak();  // Output: "Animal speaks"
//     dog.bark();   // Output: "Dog barks"
//     cat.speak();  // Output: "Animal speaks"
//     cat.meow();   // Output: "Cat meows"
//     return 0;
// }



////Hybrid inheritance
// #include <bits/stdc++.h>
// using namespace std;

// class Animal {
// public:
//     void speak() {
//         cout << "Animal speaks" << endl;
//     }
// };

// class Mammal : public Animal {
// public:
//     void run() {
//         cout << "Mammal runs" << endl;
//     }
// };

// class Bird {
// public:
//     void fly() {
//         cout << "Bird flies" << endl;
//     }
// };

// class Bat : public Mammal, public Bird {
// public:
//     void navigate() {
//         cout << "Bat navigates in the dark" << endl;
//     }
// };

// int main() {
//     Bat bat;
//     bat.speak();   // Output: "Animal speaks"
//     bat.run();     // Output: "Mammal runs"
//     bat.fly();     // Output: "Bird flies"
//     bat.navigate(); // Output: "Bat navigates in the dark"
//     return 0;
// }



//--------------------------------Polymorphism----------------------------
///simple 
// #include <bits/stdc++.h>
// using namespace std;

// class Shape {
// public:
//     virtual void draw() {
//         cout << "Drawing a shape" << endl;
//     }
// };

// class Circle : public Shape {
// public:
//     void draw() override {
//         cout << "Drawing a circle" << endl;
//     }
// };

// class Square : public Shape {
// public:
//     void draw() override {
//         cout << "Drawing a square" << endl;
//     }
// };

// int main() {
//     Shape* shape1 = new Circle();
//     Shape* shape2 = new Square();

//     shape1->draw();  // Output: Drawing a circle
//     shape2->draw();  // Output: Drawing a square

//     delete shape1;
//     delete shape2;

//     return 0;
// }



////Method Overriding 
// #include <bits/stdc++.h>
// using namespace std;

// class Animal {
// public:
//     virtual void sound() {
//         cout << "Animal makes a sound" << endl;
//     }
// };

// class Dog : public Animal {
// public:
//     void sound() override {
//         cout << "Dog barks" << endl;
//     }
// };

// int main() {
//     Animal animal;
//     Dog dog;

//     animal.sound();  // Output: Animal makes a sound
//     dog.sound();     // Output: Dog barks

//     return 0;
// }



////Method Overloading 
// #include <bits/stdc++.h>
// using namespace std;

// class Calculator {
// public:
//     int add(int a) {
//         return a;
//     }

//     int add(int a, int b) {
//         return a + b;
//     }

//     int add(int a, int b, int c) {
//         return a + b + c;
//     }
// };

// int main() {
//     Calculator calc;

//     cout << calc.add(1) << endl;           // Output: 1
//     cout << calc.add(1, 2) << endl;        // Output: 3
//     cout << calc.add(1, 2, 3) << endl;     // Output: 6

//     return 0;
// }



//-------------------------------Abstration-------------------------
////simple
// #include <bits/stdc++.h>
// using namespace std;
// // Abstract class declaration
// class AbstractShape {
// public:
//     // Pure virtual function (abstract method)
//     virtual void draw() const = 0;

//     // Normal member function
//     void displayArea() const {
//         cout << "Area calculation not available for this shape." << endl;
//     }

//     // Virtual destructor for polymorphic behavior
//     virtual ~AbstractShape() {}
// };

// // Concrete subclass of the abstract class
// class Circle : public AbstractShape {
// public:
//     void draw() const override {
//         cout << "Drawing a circle." << endl;
//     }
// };

// // Concrete subclass of the abstract class
// class Square : public AbstractShape {
// public:
//     void draw() const override {
//         cout << "Drawing a square." << endl;
//     }
// };

// // Usage of abstract class and concrete subclasses
// int main() {
//     Circle circle;
//     Square square;

//     circle.draw();        // Output: Drawing a circle.
//     circle.displayArea(); // Output: Area calculation not available for this shape.

//     square.draw();        // Output: Drawing a square.
//     square.displayArea(); // Output: Area calculation not available for this shape.

//     return 0;
// }



////interface
// #include <bits/stdc++.h>
//using namespace std;
// // Interface (Abstract class) declaration
// class Shape {
// public:
//     virtual ~Shape() {}  // Virtual destructor for polymorphic behavior
//     virtual double calculateArea() const = 0;  // Pure virtual function (interface method)
// };

// // Class implementing the interface
// class Circle : public Shape {
// private:
//     double radius;

// public:
//     Circle(double r) : radius(r) {}

//     double calculateArea() const override {
//         return 3.14 * radius * radius;
//     }
// };

// // Class implementing the interface
// class Square : public Shape {
// private:
//     double sideLength;

// public:
//     Square(double length) : sideLength(length) {}

//     double calculateArea() const override {
//         return sideLength * sideLength;
//     }
// };

// // Usage of interface and concrete classes
// int main() {
//     Circle circle(5);
//     Square square(4);

//     cout << "Area of Circle: " << circle.calculateArea() << endl;  // Output: Area of Circle: 78.5
//     cout << "Area of Square: " << square.calculateArea() << endl;  // Output: Area of Square: 16

//     return 0;
// }


////Abstract Classes
// #include <bits/stdc++.h>
//using namespace std;
// // Abstract class declaration
// class Animal {
// public:
//     virtual ~Animal() {}  // Virtual destructor for polymorphic behavior
//     virtual void makeSound() const = 0;  // Pure virtual function (abstract method)

//     void sleep() const {
//         cout << "Zzz..." << endl;
//     }
// };

// // Concrete subclass of the abstract class
// class Dog : public Animal {
// public:
//     void makeSound() const override {
//         cout << "Woof!" << endl;
//     }
// };

// // Concrete subclass of the abstract class
// class Cat : public Animal {
// public:
//     void makeSound() const override {
//         cout << "Meow!" << endl;
//     }
// };

// // Usage of abstract class and concrete subclasses
// int main() {
//     Dog dog;
//     Cat cat;

//     dog.makeSound();  // Output: Woof!
//     dog.sleep();      // Output: Zzz...

//     cat.makeSound();  // Output: Meow!
//     cat.sleep();      // Output: Zzz...

//     return 0;
// }
